library/core/src/option.rs

```diff
@@ -72,16 +72,19 @@
 //!
 //! # Representation
 //!
-//! Rust guarantees to optimise the following inner types such that an [`Option`] which contains
-//! them has the same size as a pointer:
+//! Rust guarantees to optimize the following types `<T>` such that an
+//! [`Option<T>`] has the same size as `T`:
 //!
+//! * [`Box<T>`]
 //! * `&T`
 //! * `&mut T`
 //! * `extern "C" fn`
 //! * [`num::NonZero*`]
 //! * [`ptr::NonNull<T>`]
 //! * `#[repr(transparent)]` struct around one of the types in this list.
-//! * [`Box<T>`]
+//!
+//! For the above cases, it is guaranteed that one can use [`mem::transmute`]
+//! between `T` and `Option<T>` and vice versa.
```

Comment:
No, not vice versa! If you transmute a `None` from `Option<T>` to `T`, you get UB.

Reply:
Good point. Should there be a guarantee for non-`None` in such a case, or just rule out the reverse direction entirely?

Reply:
Guarantee for non-`None` seems reasonable.

Reply:
Fixed in https://github.com/rust-lang/rust/pull/75454/commits/83f47aa11bd664ed8a15ef9833063833b7b3e71c.